samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/servlet/cp/DownloadCPServlet.java

```diff
@@ -79,34 +85,139 @@ public void doGet(HttpServletRequest req, HttpServletResponse res)
+			int success = assessmentService.updateAllRandomPoolQuestions(assessmentService.getAssessment(assessmentId));
+			if(success == AssessmentService.UPDATE_SUCCESS) {
+				currentSiteId = assessmentService.getAssessmentSiteId(assessmentId);
+				createdBy = assessmentService.getAssessmentCreatedBy(assessmentId);
+				if (canExport(req, res, agentIdString, currentSiteId, createdBy)) {
+					accessDenied = false;
+				}
+
+				if (accessDenied) {
+					String path = "/jsf/qti/exportDenied.faces";
+					RequestDispatcher dispatcher = req.getRequestDispatcher(path);
+					dispatcher.forward(req, res);
+				} else {
+					res.setContentType("application/x-zip-compressed");
+					String zipFilename = "exportAssessment.zip";
+					res.setHeader("Content-Disposition", "attachment;filename=\"" + zipFilename + "\";");
+
+					ServletOutputStream outputStream = null;
+					ZipOutputStream zos = null;
+					ZipEntry ze = null;
+
+					try {
+						byte[] b = null;
+						outputStream = res.getOutputStream();
+						zos = new ZipOutputStream(outputStream);
+
+						// QTI file
+						ze = new ZipEntry("exportAssessment.xml");
+						zos.putNextEntry(ze);
+						XMLController xmlController = (XMLController) ContextUtil.lookupBeanFromExternalServlet("xmlController", req, res);
+						xmlController.setId(assessmentId);
+						xmlController.setQtiVersion(1);
+						xmlController.displayAssessmentXml();
+						String qtiString = xmlController.getXmlBean().getXml();
+						log.debug("qtiString = " + qtiString);
+						b = qtiString.getBytes();
+						zos.write(b, 0, b.length);
+						zos.closeEntry();
+
+						// imsmanifest.xml
+						ze = new ZipEntry("imsmanifest.xml");
+						zos.putNextEntry(ze);
+						ManifestGenerator manifestGenerator = new ManifestGenerator(assessmentId);
+						String manString = manifestGenerator.getManifest();
+						log.debug("manString = " + manString);
+						b = manString.getBytes();
+						zos.write(b, 0, b.length);
+						zos.closeEntry();
+
+						// Attachments
+						HashMap contentMap = manifestGenerator.getContentMap();
+
+						String filename = null;
+						for (Iterator it = contentMap.entrySet().iterator(); it.hasNext();) {
+							Map.Entry entry = (Map.Entry) it.next();
+							filename = (String)  entry.getKey();
+							ze = new ZipEntry(filename.substring(1));
+							zos.putNextEntry(ze);
+							b = (byte[]) entry.getValue();
+							zos.write(b, 0, b.length);
+							zos.closeEntry();
+						}
+
+					} catch (IOException e) {
+						log.error(e.getMessage());
+						throw e;
+					} finally {
+						if (zos != null) {
+							try {
+								zos.closeEntry();
+							} catch (IOException e) {
+								log.error(e.getMessage());
+							}
+							try {
+								zos.close();
+							} catch (IOException e) {
+								log.error(e.getMessage());
+							}
+
+						}
+					}
+				}
+			} else {
+				if (success == AssessmentService.UPDATE_ERROR_DRAW_SIZE_TOO_LARGE) {
```

Comment:
Let's use a constant for "org.sakaiproject.tool.assessment.bundle.AuthorMessages"